rfc/ipni-w3c.md

```diff
@@ -0,0 +1,346 @@
+
+See: [Encoding the IPNI Advertisement](#encoding-the-ipni-advertisement)
+
+The Advertisement should then be available for consumption by indexer nodes per the [Advertisement Transfer](https://github.com/ipni/specs/blob/main/IPNI.md#advertisement-transfer) section of the IPNI spec.
+
+The w3s service is responsible for creating advertisements. Having users publish the advertisements is not practical as that would require the user to maintain an advertisement chain and serve a network endpoint from which indexers can fetch the advertisements.
+
+#### Encoding the IPNI Advertisement
+
+The set of multihashes within one CAR must be encoded as 1 Advertisement, having 1 or more `EntryChunk` blocks. `EntryChunk` blocks include an array of multihashes.
+
+The `EntryChunk` serves a similar purpose as [HTTP chunking](https://en.wikipedia.org/wiki/Chunked_transfer_encoding). A typical block will have 16k multihashes in it.
+
+A CARx-index encodes a set of multihashes. Mapping from an index to an `EntryChunk` requires parsing the index and encoding the multihashes it contains with the EntryChunk IPLD schema.
+
+```ipldsch
+type EntryChunk struct {
+    Entries [Bytes]
+    Next optional Link
+}
+```
+
+Where the IPLD encoded size of an `EntryChunk` with the set of multihashes would exceed 4MiB (the upper limit for a block that can be transferred by libp2p) the set of multihashes must be split into multiple `EntryChunk` blocks.
+
+All multihashes must be sorted in ascending order before being split into chunks. This is done to have deterministic output when creating entry chinks - the same set of multihashes should generate the same set of entry chunks.
+
+### Finding and retrieving stored data
+
+```mermaid
+sequenceDiagram
+    actor Alice
+    Alice->>ipni: query (CID)
+    activate ipni
+    ipni-->>Alice: claims location, CAR CID
+    deactivate ipni
+    Alice->>w3s: get claims (URL)
+    activate w3s
+    w3s-->>Alice: claims
+    deactivate w3s
+    activate Alice
+    Alice-->>Alice: read CARx-index location
+    Alice->>w3s: get CARx-index (URL)
+    deactivate Alice
+    activate w3s
+    w3s-->>Alice: CARx-index
+    deactivate w3s
+    activate Alice
+    Alice-->>Alice: read offset/range for CID(s)
+    Alice->>w3s: get data (CAR URL, data CID, offset, len)
+    deactivate Alice
+    activate w3s
+    w3s-->>w3s: fetch CAR
+    w3s-->>w3s: extract data from CAR
+    w3s-->>Alice: data item(s)
+    deactivate w3s
+```
+        
+**Note**: One round trip can be avoided if the CARx-index location is in the IPNI response, instead of having to read it out of a location claim. This may not work if the CARx-index location claim is too big for the metadata (recommended size: 100 bytes, max size: 1024 bytes).
+
+**Question**: 
+Will the claims bucket be located at some arbitrary URL, or is it stored in a location known to w3s? If the latter then only the CID of the cliams bucket or CID of the CAR file is needed to fetch claims.
```

Comment:
Let's assume for the moment that the retriever is a public client, and THEY don't know the location absent a public URL, even if W3S does.

Reply:
OK, This makes sense to encode into the `Addresses` field of an advertisement. This allows there to be multiple addresses that can easily be changed by updating in a more recent advertisement. That address plus the cliaims bucket CID from the metadata will give us what we need.